File: backend/app/schemas/geo.py
# =============================================================================
# FGA CRM - Schemas GEO (Generative Engine Optimization)
# =============================================================================
"""Schemas Pydantic v2 du module GEO.

- Enums : GeoEngine, GeoIntent, GeoSentiment
- Create / Update / Response pour chaque modele
- Schemas d'extraction (interne, non exposes en API)
- Schemas API speciaux (trigger, dashboard, health)
"""

from datetime import date, datetime
from enum import StrEnum
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------
# StrEnum (py3.11+) : equivalent fonctionnel a (str, Enum) — .value et la
# comparaison avec un str fonctionnent a l'identique. Choisi pour rester
# ruff-compatible (regle UP042) sur ce projet cible py312.

class GeoEngine(StrEnum):
    perplexity = "perplexity"
    openai = "openai"
    gemini = "gemini"
    claude = "claude"
    grok = "grok"
    google_aio = "google_aio"


class GeoIntent(StrEnum):
    informationnel = "informationnel"
    comparatif = "comparatif"
    transactionnel = "transactionnel"


class GeoSentiment(StrEnum):
    positif = "positif"
    neutre = "neutre"
    negatif = "negatif"


# ---------------------------------------------------------------------------
# GeoBrand
# ---------------------------------------------------------------------------

class GeoBrandCreate(BaseModel):
    organization_id: UUID | None = None
    slug: str = Field(..., min_length=1, max_length=255)
    name: str = Field(..., min_length=1, max_length=255)
    aliases: list[Annotated[str, Field(max_length=255)]] = Field(default_factory=list, max_length=50)
    is_owned: bool = False
    active: bool = True


class GeoBrandUpdate(BaseModel):
    organization_id: UUID | None = None
    slug: str | None = Field(None, min_length=1, max_length=255)
    name: str | None = Field(None, min_length=1, max_length=255)
    aliases: list[Annotated[str, Field(max_length=255)]] | None = Field(None, max_length=50)
    is_owned: bool | None = None
    active: bool | None = None


class GeoBrandResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID | None
    slug: str
    name: str
    aliases: list[str]
    is_owned: bool
    active: bool
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# GeoPrompt
# ---------------------------------------------------------------------------

class GeoPromptCreate(BaseModel):
    organization_id: UUID | None = None
    text: str = Field(..., min_length=1, max_length=2000)
    intent: GeoIntent
    persona: str | None = Field(None, max_length=255)
    country: str = Field("FR", max_length=8)
    language: str = Field("fr", max_length=8)
    tags: list[str] = Field(default_factory=list, max_length=30)
    priority: bool = False
    active: bool = True


class GeoPromptUpdate(BaseModel):
    text: str | None = Field(None, min_length=1, max_length=2000)
    intent: GeoIntent | None = None
    persona: str | None = Field(None, max_length=255)
    country: str | None = Field(None, max_length=8)
    language: str | None = Field(None, max_length=8)
    tags: list[str] | None = Field(None, max_length=30)
    priority: bool | None = None
    active: bool | None = None


class GeoPromptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID | None
    brand_id: UUID
    text: str
    intent: str
    persona: str | None
    country: str
    language: str
    tags: list[str]
    priority: bool
    active: bool
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# GeoRun
# ---------------------------------------------------------------------------

class GeoRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    prompt_id: UUID
    brand_id: UUID
    run_index: int
    engine: str
    model_version: str | None
    country: str | None
    language: str | None
    run_at: datetime
    raw_answer: str | None
    citations: list[dict]
    brands_found: list[dict]
    brand_mentioned: bool | None
    brand_position: int | None
    brand_sentiment: str | None
    brand_recommended: bool | None
    appearance: bool | None
    created_at: datetime


class GeoRunListResponse(BaseModel):
    items: list[GeoRunResponse]
    total: int
    page: int
    size: int
    pages: int


# ---------------------------------------------------------------------------
# GeoMetricsDaily
# ---------------------------------------------------------------------------

class GeoMetricsDailyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID | None
    day: date
    brand_id: UUID
    engine: str
    visibility_rate: float | None
    sov: float | None
    sov_weighted: float | None
    sentiment_avg: float | None
    reco_rate: float | None
    runs_total: int
    computed_at: datetime


# ---------------------------------------------------------------------------
# Extraction (interne — non expose en API)
# ---------------------------------------------------------------------------

class MarqueTrouvee(BaseModel):
    nom: str
    rang: int = Field(ge=1)
    recommandee: bool
    sentiment: GeoSentiment
    justification: str = Field(max_length=500)


class ExtractionResult(BaseModel):
    marques: list[MarqueTrouvee]


# ---------------------------------------------------------------------------
# Schemas API speciaux
# ---------------------------------------------------------------------------

class GeoRunTriggerRequest(BaseModel):
    brand_id: UUID
    engine: GeoEngine
    prompt_ids: list[UUID] = Field(min_length=1, max_length=50)
    n_runs: int = Field(default=3, ge=1, le=5)
    country: str = Field("FR", max_length=8)
    language: str = Field("fr", max_length=8)


class GeoRunTriggerResponse(BaseModel):
    task_id: str
    runs_scheduled: int


class GeoDashboardResponse(BaseModel):
    brand: GeoBrandResponse
    engine: str
    date_from: date
    date_to: date
    metrics: list[GeoMetricsDailyResponse]
    top_competitors: list[dict]   # [{nom, mentions, sov_share}]
    top_sources: list[dict]       # [{domain, count}]


class GeoHealthResponse(BaseModel):
    engine: str
    status: str   # ok|error|unconfigured
    checked_at: datetime
    error: str | None = None

File: frontend/src/types/geo.ts
// =============================================================================
// FGA CRM - Types GEO (Generative Engine Optimization)
// =============================================================================
// Types alignes EXACTEMENT sur les schemas backend (app/schemas/geo.py) — DC10.
// Toute divergence de nom de champ = source de bug silencieux cote front.
// =============================================================================

export type GeoEngine = 'perplexity' | 'openai' | 'gemini' | 'claude' | 'grok' | 'google_aio';
export type GeoIntent = 'informationnel' | 'comparatif' | 'transactionnel';
export type GeoSentiment = 'positif' | 'neutre' | 'negatif';

// GeoBrandResponse
export interface GeoBrand {
  id: string;
  slug: string;
  name: string;
  aliases: string[];
  is_owned: boolean;
  active: boolean;
  created_at: string;
  updated_at: string;
}

// GeoPromptResponse — note : intent renvoye en str par le backend
export interface GeoPrompt {
  id: string;
  brand_id: string;
  text: string;
  intent: GeoIntent;
  persona: string | null;
  country: string;
  language: string;
  tags: string[];
  priority: boolean;
  active: boolean;
  created_at: string;
  updated_at: string;
}

// GeoBrandOverviewResponse — marque + visibilite moyenne (selecteur avec mini-score)
export interface GeoBrandOverview {
  id: string;
  slug: string;
  name: string;
  visibility_rate: number | null;
}

// Payloads de creation (alignes sur GeoBrandCreate / GeoPromptCreate backend)
export interface GeoBrandInput {
  slug: string;
  name: string;
  aliases: string[];
  is_owned: boolean;
  active?: boolean;
}

export interface GeoPromptInput {
  text: string;
  intent: GeoIntent;
  persona?: string | null;
  country?: string;
  language?: string;
  tags?: string[];
  priority?: boolean;
  active?: boolean;
}

// GeoRunResponse — backend expose raw_answer + created_at (cf schemas/geo.py)
export interface GeoRun {
  id: string;
  prompt_id: string;
  brand_id: string;
  run_index: number;
  engine: GeoEngine;
  model_version: string | null;
  country: string | null;
  language: string | null;
  run_at: string;
  raw_answer: string | null;
  citations: Array<{ url: string; domain: string; rank: number }>;
  brands_found: Array<{ nom: string; rang: number; recommandee: boolean; sentiment: GeoSentiment }>;
  brand_mentioned: boolean | null;
  brand_position: number | null;
  brand_sentiment: GeoSentiment | null;
  brand_recommended: boolean | null;
  appearance: boolean | null;
  created_at: string;
}

// GeoMetricsDailyResponse — backend expose computed_at en plus
export interface GeoMetricsDaily {
  id: string;
  day: string;
  brand_id: string;
  engine: GeoEngine;
  visibility_rate: number | null;
  sov: number | null;
  sov_weighted: number | null;
  sentiment_avg: number | null;
  reco_rate: number | null;
  runs_total: number;
  computed_at: string;
}

// GeoDashboardResponse
export interface GeoDashboard {
  brand: GeoBrand;
  engine: string;
  date_from: string;
  date_to: string;
  metrics: GeoMetricsDaily[];
  top_competitors: Array<{ nom: string; mentions: number; sov_share: number }>;
  top_sources: Array<{ domain: string; count: number }>;
}

// Reponse de GET /brands/{id}/gaps (dicts bruts cote backend — pas de schema Pydantic)
export interface GeoGap {
  prompt_id: string;
  prompt_text: string;
  intent: GeoIntent;
  priority: boolean;
  runs_checked: number;
  mentions: number;
  visibility_rate: number;
  top_competitor_sources: Array<{ domain: string; count: number }>;
  top_competitors: Array<{ nom: string; count: number }>;
  last_run_at: string | null;
  action_suggestion: string;
}

// Reponse de GET /brands/{id}/alerts (dicts bruts cote backend)
export interface GeoAlert {
  alert_type: 'sov_drop' | 'visibility_zero' | 'sentiment_negative' | 'competitor_overtake';
  engine: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  detail: Record<string, unknown>;
  detected_at: string;
}

// GeoHealthResponse
export interface GeoHealth {
  engine: string;
  status: 'ok' | 'error' | 'unconfigured';
  checked_at: string;
  error: string | null;
}

// GeoRunTriggerResponse
export interface GeoRunTriggerResponse {
  task_id: string;
  runs_scheduled: number;
}

// GeoRunListResponse
export interface GeoRunsPage {
  items: GeoRun[];
  total: number;
  page: number;
  size: number;
  pages: number;
}
